backend/src/api/model/block/mutations.rs

```diff
@@ -0,0 +1,355 @@
+                    "update blocks
+                        set index = updates.new_index
+                        from realms, (values
+                            ($1::smallint, $2::smallint),
+                            ($2::smallint, $1::smallint)
+                        ) as updates(old_index, new_index), (
+                            select count(*) as count from blocks
+                            where realm_id = $3
+                        ) as count
+                        where realm_id = realms.id
+                        and realm_id = $3
+                        and blocks.index = updates.old_index
+                        and updates.new_index < count
+                        and updates.new_index >= 0
+                        returning {}",
+                    Realm::col_names("realms"),
+                ),
+                dbargs![
+                    &(index_1 as i16),
+                    &(index_2 as i16),
+                    &realm.key_for(Id::REALM_KIND)
+                        .ok_or_else(|| invalid_input!("`realm` is not a valid realm id"))?,
+                ],
+            )
+            .await?
+            // We will get the realm twice for two updated rows
+            // if everything goes according to plan.
+            // If we skip the first and can successfully grab the second,
+            // we know it did.
+            .skip(1);
+
+        pin_mut!(realm_stream);
+
+        let realm = realm_stream
+            .next()
+            .await
+            .ok_or_else(|| invalid_input!(
+                "`index1`, `index2` or `realm` wasn't a valid block index or realm"
+            ))?
+            .map(Realm::from_row)?;
+
+        Ok(realm)
+    }
+
+    pub(crate) async fn update(id: Id, set: UpdateBlock, context: &Context) -> ApiResult<BlockValue> {
+        let updated_block = context.db(context.require_moderator()?)
+            .query_one(
+                &format!(
+                    "update blocks set
+                        title = case $2::boolean when true then $3 else title end
+                        where id = $1
+                        returning {}",
+                    Self::COL_NAMES,
+                ),
+                &[
+                    &id.key_for(Id::BLOCK_KIND)
+                        .ok_or_else(|| invalid_input!("`id` does not refer to a block"))?,
+                    &!set.title.is_implicit_null(),
+                    &set.title.some(),
+                ],
+            )
+            .await?;
+
+        Ok(Self::from_row(updated_block)?)
+    }
+
+    pub(crate) async fn update_text(id: Id, set: UpdateTextBlock, context: &Context) -> ApiResult<BlockValue> {
+        let updated_block = context.db(context.require_moderator()?)
+            .query_one(
+                &format!(
+                    "update blocks set
+                        title = case $2::boolean when true then $3 else title end,
+                        text_content = coalesce($4, title)
```

Comment:
Yes regarding the `text_content`.

Regarding the second point, there is a difference. I wanted the ability to explicitly set the title to `null` again, since the frontend actually renders `""` titles differently, IIRC. However, that's not possible with `coalesce`. If you only ever update a field with `coalesce(argument, field)`, once the field was non-null once, that expression never evaluate to `null` again.

Reply:
Ah ok makes sense.